File: python_backend/id_assign_/top/id_assign.py
from id_assign_.sub_ids import sub_ids
from id_assign_.rent_ids import rent_ids
from id_assign_.get_standard_ids import get_standard_ids
from pprint import pprint


def id_assign(custom_piece_names, subs):
    """
    """
    id_dict = sub_ids(subs)
    id_dict = rent_ids(id_dict, custom_piece_names, ignore=subs.keys())  # ignore because taken care of previous step.
    id_dict.update(get_standard_ids(exclude=subs.values())) # exclude standard pieces subed for.
    return id_dict


if __name__ == "__main__":
    pprint(id_assign(['Jester',
                      'Duke',
                      'Joker',
                      'Morty'],
                     {'Jester': 'Queen', 'Duke': 'Bishop'}))

File: app/src/apiHelpers/idAssign/top/idAssign.js
import {rentIds} from "../rentIds";
import {getStandardIds} from "../getStandardIds";
import {subIds} from "../subIds";

export function idAssign(customPieceNames, subs) {
    /**
    assign custom piece names to standard piece ids, therefore "subbing" for those standard pieces. Standard ids not subed
    for keep their standard piece assignments. For custom pieces not subed but that are pawn promotion options, rent from
    1 of 20 non-standard ids.
    */
    var idDict = subIds(subs)
    idDict = rentIds(idDict, customPieceNames, Object.keys(subs))  // "ignore" subs because just did that
    idDict = {...idDict, ...getStandardIds(Object.values(subs)) } // get ids the of standard pieces not subbing for.
    return idDict;
}


/**  for node.js
if (require.main === module) {
    console.log(idAssign(['Jester',
                      'Duke',
                      'Joker',
                      'Morty'], {'Jester': 'Queen', 'Duke': 'Bishop'}))
}
*/